Azure Monitor cli commands for Managemnt & Data plane

vishrutshah opened the pull request:
This enables support for #1954 

Here is the complete set of available commands exposed from [azure-mgmt-monitor](https://github.com/Azure/azure-sdk-for-python/tree/master/azure-mgmt-monitor) and [azure-monitor](https://github.com/Azure/azure-sdk-for-python/tree/master/azure-monitor) python sdk. 

Here is the state of each Management command:

|  | Command  | Tested [Manual]  | Status  | Issue  | Category|
|---|---|---|---|---|---|
| log-profiles |   |   |   |   |
|  | create  |   |   | [Issue 973](https://github.com/Azure/azure-sdk-for-python/issues/973)  | Swagger |
|  | delete | :white_check_mark:  | :white_check_mark:  |   |
|  | show  | :white_check_mark:  | :white_check_mark:  |   |
|  | list  | :white_check_mark:  | :white_check_mark:  |  |
|  | update  |   |   | [ Issue 973](https://github.com/Azure/azure-sdk-for-python/issues/973) |Swagger |
|---|---|---|---|---|
| alert-rule-incidents |   |   |   |   |
|  | show  | :white_check_mark:  | :white_check_mark:  |   |
|  | list  | :white_check_mark:  | :white_check_mark:  |  |
|---|---|---|---|---|
| alert-rules  |   |   |   |   |
|  | create  |   |   | [Issue 971](https://github.com/Azure/azure-sdk-for-python/issues/971) [Issue 11](https://github.com/Azure/msrest-for-python/issues/11)  | AutoRest Python-Runtime |
|  | delete | :white_check_mark:  | :white_check_mark:  |   |
|  | show  |   |   | [Issue 971](https://github.com/Azure/azure-sdk-for-python/issues/971)  | AutoRest|
|  | list  |  |   |  [Issue 970](https://github.com/Azure/azure-sdk-for-python/issues/970)  | AutoRest|
|  | update  |   |   |[Issue 971](https://github.com/Azure/azure-sdk-for-python/issues/971) [Issue 11](https://github.com/Azure/msrest-for-python/issues/11) | AutoRest Python-Runtime |
|---|---|---|---|---|
| service-diagnostic-settings  |   |   |   |   |
|  | create  |   |   | [Issue 972](https://github.com/Azure/azure-sdk-for-python/issues/972)  | Swagger |
|  | show  | :white_check_mark:  | :white_check_mark:  |   |
|  | update|   |   | [Issue 972](https://github.com/Azure/azure-sdk-for-python/issues/972)  | Swagger |
|---|---|---|---|---|
| autoscale-settings  |   |   |   | Known Issue in Service  | Monitor Service |
|  | create  | No  |   |  |
|  | delete |  No |   |   |
|  | show  | No |   |   |
|  | list  | No  |  |  | 
|  | update  |  No |   |  |

Here is the state of each Data Plane command:

|  | Command  | Tested [Manual]  | Status  | Issue  | Category |
|---|---|---|---|---|---|
| usage-metrics |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |
| event-categories |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |
| events |   |   |   |   | |
|  | list  |  |  | [Issue 969](https://github.com/Azure/azure-sdk-for-python/issues/969) | Swagger |
| tenant-events |   |   |   |   | |
|  | list  |  |  | [Issue 899](https://github.com/Azure/azure-rest-api-specs/issues/899) | Permission |
| metric-definitions |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |
| metrics |   |   |   |   | |
|  | list  | :white_check_mark:  | :white_check_mark:  |  | |

**Note:** [azure-mgmt-monitor](https://github.com/Azure/azure-sdk-for-python/tree/master/azure-mgmt-monitor) is not released as python package yet, once released we'll take our dependency on locally checked-in sdk before merge. 

@troydai @derekbekoe & @tjprescott when you get a chance, please have a first quick look. Thanks

vishrutshah commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py:
```diff
@@ -0,0 +1,194 @@
+
+    def expand(self, argument_name, model_type, group_name=None, patches=None):
+        # TODO:
+        # two privates symbols are imported here. they should be made public or this utility class
```
I think this would be a good time to move this utility class into azure.cli.core

tjprescott replied:
I disagree.  If done, it should be a separate PR to not make this review more difficult.

vishrutshah replied:
Makes sense. Then I'd prefer it to refactor once this PR gets in.

tjprescott commented:
You need to fix the linter issuer (add "# pylint: skip-file" to any auto-genned SDK) and also add a comment to the PR with the -h output for the most important commands.

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py:
```diff
@@ -0,0 +1,194 @@
+from azure.cli.core.commands.arm import cli_generic_update_command
+
+
+# MANAGEMENT CLIENT FACTORIES
```
These should go in a _client_factory.py file to be consistent with other packages.

derekbekoe replied:
Still think the client factories should be in _client_factory.py file.

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py:
```diff
@@ -0,0 +1,74 @@
+            """
+helps['monitor log-profiles update'] = """
+            type: command
+            short-summary: Update a log profile in Azure Monitoring REST API.
```
No other command mentions a "REST API" in the CLI that I know of.

vishrutshah replied:
Ya that's true..that came from swagger comment...I'll remove it 👍

vishrutshah replied:
Updated: Update a log profile assigned to Azure subscription

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py:
```diff
@@ -0,0 +1,74 @@
+            """
+helps['monitor usage-metrics'] = """
+            type: group
+            short-summary: Commands to manage usage matrics.
```
typo: metrics.

vishrutshah replied:
Fixed

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/help.py:
```diff
@@ -0,0 +1,74 @@
+# MANAGEMENT COMMANDS HELPS
+helps['monitor alert-rules'] = """
+            type: group
+            short-summary: Commands to manage alert rules.
```
It might make sense to clarify "manage Monitor alert rules" etc. for these commands since the terms are pretty generic and could be confused for other things.

vishrutshah replied:
Updated: Commands to manage alerts assigned to Azure resources

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py:
```diff
@@ -0,0 +1,47 @@
+
+with ParametersContext(command='monitor alert-rules') as c:
+    c.register_alias('name', ('--azure-resource-name',))
+    c.register_alias('rule_name', ('--name', '-n'))
```
The `--name` parameters here need a values completer within their scope (except for create commands).

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py:
```diff
@@ -0,0 +1,47 @@
+    c.register_alias('log_profile_name', ('--name', '-n'))
+
+with ParametersContext(command='monitor autoscale-settings') as c:
+    c.register_alias('name', ('--azure-resource-name',))
```
`--azure-resource-name` is pretty verbose. Also, will this accept a name or id?  Should it?  If so, I would recommend `--resource` with help text that says "Name or ID of the resource to query." or something to that effect.

vishrutshah replied:
I haven't done tests on autoscale, so i am not sure on whether it can accept ids. So I'll keep this pending until then :)

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/params.py:
```diff
@@ -0,0 +1,47 @@
+    c.expand('parameters', AlertRuleResource)
+    c.register_type('condition', ('--condition',),
+                    json.loads,
+                    'JSON encoded condition configuration. Use @{file} to load from a file.')
```
I recommend you spell out the kwargs like we do in the traditional model. (i.e. type=json.loads, help='blahblah')

vishrutshah replied:
Cool

tjprescott commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/validators.py:
```diff
@@ -0,0 +1,23 @@
+    """
+    Return a validator which will aggregate multiple arguments to one complex argument.
+    """
+    def _expension_valiator_impl(namespace):
```
Typo: expansion

derekbekoe replied:
typo: validator

vishrutshah replied:
fixed

derekbekoe commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/__init__.py:
```diff
@@ -0,0 +1,16 @@
+
+import azure.cli.command_modules.monitor.help  # pylint: disable=unused-import
+
+__all__ = ['load_params', 'load_commands']
```
What's this line used for?
It shouldn't be needed.

troydai replied:
It's just a formality though it is not technically needed.

vishrutshah replied:
removed

derekbekoe commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/commands.py:
```diff
@@ -0,0 +1,119 @@
+# MANAGEMENT COMMANDS
+alert_rules_operations = create_service_adapter(
+    'azure.cli.command_modules.monitor.sdk.operations.alert_rules_operations',
+    'AlertRulesOperations')
```
This is another way of just doing `modpath.format(a, b, c)` like in the VM module (https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/commands.py#L19).

I don't really see this implementation as *simpler* than what's in the VM module as I now have to follow `create_service_adapter` to see what that does.

vishrutshah replied:
I see but as I am following the flow/structure from `azure-cli-sql` (as I have understood that flow as of now :)), I'd prefer to keep it same. What's your thought on that?

derekbekoe commented on src/command_modules/azure-cli-monitor/setup.py:
```diff
@@ -0,0 +1,55 @@
+DEPENDENCIES = [
+    'azure-cli-core',
+    'azure-monitor==0.1.0'
+    # 'azure-mgmt-monitor==0.1.0'
```
Please remove if it's not used.

vishrutshah replied:
@derekbekoe I was thinking that we'll not merge the PR in as it has sdk checked-in. So I was planning to wait until azure-mgmt-monitor releases and then removing sdks and the merging PR. 

Let me know if there is way where we still allow merging PRs if sdks are checked-in? If so i'll definitely remove it.

troydai replied:
No. We should NOT. No matter If azure-mgmt-monitor is released we should not include their source code in our code base.

derekbekoe replied:
Okay didn't know you weren't planning on merging this right now.
I've created and added a Do Not Merge label to this PR then.

tjprescott replied:
We have allowed services (KeyVault, IoT, etc.) to merge PRs with embedded clients. DocumentDB was about to earlier this week but the SDK was released so he went ahead and did the import changes.

If something holds up this PR from being merged, it really doesn't need to be the desire to avoid (temporarily) having the SDK code in the module.

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor -h

Group
    az monitor: Commands to manage Azure Monitor service.

Subgroups:
    alert-rule-incidents       : Commands to manage alert rule incidents.
    alert-rules                : Commands to manage alert rules.
    autoscale-settings         : Commands to manage autoscale settings.
    event-categories           : Commands to manage event categories.
    events                     : Commands to manage events.
    log-profiles               : Commands to manage log profiles.
    metric-definitions         : Commands to manage metric definitions.
    metrics                    : Commands to manage metrics.
    service-diagnostic-settings: Commands to manage service diagnostic settings.
    tenant-events              : Commands to manage tenant events.
    usage-metrics              : Commands to manage usage matrics.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor alert-rule-incidents -h

Group
    az monitor alert-rule-incidents: Commands to manage alert rule incidents.

Commands:
    list: Gets a list of incidents associated to an alert rule.
    show: Gets an incident associated to an alert rule.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor alert-rules -h

Group
    az monitor alert-rules: Commands to manage alert rules.

Commands:
    create: Creates or updates an alert rule.
    delete: Deletes an alert rule.
    list  : List the alert rules within a resource group.
    show  : Gets an alert rule.
    update: Updates an alert rule.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor autoscale-settings -h

Group
    az monitor autoscale-settings: Commands to manage autoscale settings.

Commands:
    create: Creates or updates an autoscale setting.
    delete: Deletes and autoscale setting.
    list  : Lists the autoscale settings for a resource group.
    show  : Gets an autoscale setting.
    update: Updates an autoscale setting.
```

vishrutshah commented:
```
(env) vishrut@visshamac azure-cli-monitor (monitor-cli) $ az monitor event-categories -h

Group
    az monitor event-categories: Commands to manage event categories.

Commands:
    list: Get the list of available event categories supported in the Activity Log Service.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor events -h

Group
    az monitor events: Commands to manage events.

Commands:
    list: Provides the list of events.
```

vishrutshah commented:
```
(env) vishrut@visshamac azure-cli-monitor (monitor-cli) $ az monitor log-profiles -h

Group
    az monitor log-profiles: Commands to manage log profiles.

Commands:
    create: Create or update a log profile in Azure Monitoring REST API.
    delete: Deletes the log profile.
    list  : List the log profiles.
    show  : Gets the log profile.
    update: Update a log profile in Azure Monitoring REST API.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor metric-definitions -h

Group
    az monitor metric-definitions: Commands to manage metric definitions.

Commands:
    list: Lists the metric definitions for the resource.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor metrics -h

Group
    az monitor metrics: Commands to manage metrics.

Commands:
    list: Lists the metric values for a resource.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor service-diagnostic-settings -h

Group
    az monitor service-diagnostic-settings: Commands to manage service diagnostic settings.

Commands:
    create: Create or update new diagnostic settings for the specified resource.
    show  : Gets the active diagnostic settings for the specified resource.
    update: Update diagnostic settings for the specified resource.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor tenant-events -h

Group
    az monitor tenant-events: Commands to manage tenant events.

Commands:
    list: Get the Activity Logs for the Tenant.
```

vishrutshah commented:
```
(env) vishrut@mac azure-cli-monitor (monitor-cli) $ az monitor usage-metrics -h

Group
    az monitor usage-metrics: Commands to manage usage matrics.

Commands:
    list: The List operation lists the usage metrics for the resource.
```

vishrutshah commented:
@tjprescott & @derekbekoe I have just put some more comments on showing the documentation around commands for the reference. I haven't gone through all your comments but as I go I'll keep fixing them. This is just in case you want to see the commands itself 💯 

I'll ping you guys once i fix all the review comments :)

tjprescott commented:
The help text for several of the list commands illustrates a major problem. 

First, the text from the SDK is crazily verbose, and we should replace it, but second, we should look at the equivalent Xplat CLI commands to see what convenience arguments they use to build up a filter. We should **not** simply expose the `--filter` parameter and tell people to go learn OData. Even if we do expose `--filter`, we should allow the building of a filter with argument.  See the `az resource list` command for examples of this.

troydai commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py:
```diff
@@ -0,0 +1,195 @@
+        return CommandGroup(self._scope, group_name, self._factory, self._service_adapter)
+
+
+class CommandGroup(object):
```
This is duplicate with the sql command module. Consider move it up to the `azure-cli-core` for reuse.

vishrutshah replied:
Yes, we should. I'd prefer it to do that refactoring once this PR completes so I don't confuse you guys more on actual PR. Does that seem good?

troydai commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py:
```diff
@@ -0,0 +1,195 @@
+        from azure.cli.core.commands._introspection import \
+            (extract_args_from_signature, _option_descriptions)
+
+        from .validators import get_complex_argument_processor
```
Prefer absolute path

vishrutshah replied:
Updated to azure.cli.command_modules.monitor.validators

troydai commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/sdk/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# coding=utf-8
```
are these coming from SDK? shall we commit this?

vishrutshah replied:
Nope. Will not merge in until azure-mgmt-monitor is released. PR https://github.com/Azure/azure-sdk-for-python/pull/974

vishrutshah commented:
@troydai @tjprescott @derekbekoe 

I've removed SDK code from the cli and took dependencies on `azure-mgmt-monitor` [version 0.1.0](https://pypi.python.org/pypi/azure-mgmt-monitor/0.1.0) & `azure-monitor` [version 0.2.0](https://pypi.python.org/pypi/azure-monitor/0.2.0) sdks. 

I've also addressed general code review comments except:

1. As I am following design of `azure-cli-sql`, we have some design that do not match with the other modules. Example: I do not have _client_factory.py, I follow ParametersContext to register commands, expand arguments and aliasing etc
2. I did duplicate common code between `azure-cli-sql` & `azure-monitor-cli` that will be moved and shared as separate PR
3. I did not provide very convenience methods around --filter parameters yet. I'd like to provide those convenience as separate PR
4. I've tested commands manually and currently writing automation tests around it. I can add them here or I can do that separate to make it easier for me and reviewer :)

Let me know what you guys think about new changes and how do we want to proceed.

Thanks!

derekbekoe commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/_util.py:
```diff
@@ -0,0 +1,190 @@
+def get_monitor_client(_):
+    from azure.monitor import MonitorClient
+    from azure.cli.core.commands.client_factory import get_mgmt_service_client
+    return get_mgmt_service_client(MonitorClient)
```
How does this work if you're passing in the data-plane client yet there's a call to `get_mgmt_service_client`?

vishrutshah replied:
strange...looks like i missed something here..Thanks for catching..Let me verify

vishrutshah replied:
It is indeed odd but it works :)

```
(env) vishrut@mac azure-cli (monitor-cli) $ az monitor event-categories list
[
  {
    "localizedValue": "Administrative",
    "value": "Administrative"
  },
  {
    "localizedValue": "Security",
    "value": "Security"
  },
  {
    "localizedValue": "Service Health",
    "value": "ServiceHealth"
  },
  {
    "localizedValue": "Alert",
    "value": "Alert"
  },
  {
    "localizedValue": "Recommendation",
    "value": "Recommendation"
  },
  {
    "localizedValue": "Policy",
    "value": "Policy"
  }
]
(env) vishrut@amac azure-cli (monitor-cli)
```

The reason I believe that azure-monitor & azure-mgmt-monitor sdks are generated from `Azure.Python` generator, hence the signature of client initialization is the same in both management plane and data plane sdks. So I did not ran into the any problem.

tjprescott commented:
I would not recommend merging this before the release while it exposes raw "--filter" with no helper arguments.

vishrutshah commented:
Implement filter for `metrics-definitions list` command PR: https://github.com/vishrutshah/azure-cli/pull/1. Please take a look when you get a chance. Thanks!

vishrutshah commented:
Implement filter for `metrics list` command PR: https://github.com/vishrutshah/azure-cli/pull/2. Please take a look when you get a chance. Thanks!

tjprescott commented:
@vishrutshah which commands remain that have filter building arguments in Xplat but not in this PR?

vishrutshah commented:
Implement filter for `activity-logs list` & `tenant-activity-logs list` command PR: https://github.com/vishrutshah/azure-cli/pull/3. Please take a look when you get a chance. Thanks! 

@tjprescott With this PR merged in, we have following 2 commands from _management plane_ left  to be simplified:

```
az monitor alert-rules list
```
```
az monitor autoscale-settings list
```

troydai commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py:
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
+                         timezone: 1970-01-01T00:00:00Z, 1970-01-01T00:00:00-0500
+    :param str caller: The caller to look for when querying
+    :param str status: The status value to query (ex: Failed)
+    :param str max_events: The maximum number of records to be returned by the command
```
why is it `str` not `int`?

troydai commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py:
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
+                                                      caller, status)
+
+    if max_events:
+        max_events = int(max_events)
```
can't `argparser` or validator guarantee it's integer?

troydai commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py:
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
+    '''
+    formatter = "eventTimestamp ge {} and eventTimestamp le {}"
+    odata_filters = _validate_time_range_and_add_defaults(start_time, end_time,
+                                                          formatter=formatter)
```
should time validating be done in `validator`?

troydai commented on src/command_modules/azure-cli-monitor/azure/cli/command_modules/monitor/custom.py:
```diff
@@ -97,3 +98,110 @@ def _validate_start_time(start_time, end_time):
+
+
+def _activity_logs_select_filter_builder(events=None):
+    '''Build up select filter string from events
```
use triple double quote for docstring.

tjprescott commented:
As we discussed Friday, we should close this PR until it is ready to be merged and then either reopen it or open a new one.

vishrutshah commented:
Closing this PR and sending, new one for the review for partner team.

vishrutshah commented:
New PR @ https://github.com/Azure/azure-cli/pull/2534
